src/osd/HitSet.h

```diff
@@ -450,4 +459,244 @@ class BloomHitSet : public HitSet::Impl {
 };
 WRITE_CLASS_ENCODER(BloomHitSet)
 
+/**
+ * explicitly enumerate hash hits with incremental temp in the set
+ */
+class TempHitSet : public HitSet::Impl {
+public:
+  static const uint32_t TEMPMAX = UINT32_MAX * 0.9;
+
+  template <class T>
+  struct ExclusiveKeyAsHash {
+      size_t operator()(T key) const noexcept{
+        return key;
+      }
+  };
+
+  struct temperature_t {
+    uint32_t temp;
+    uint32_t last_decay;
+
+    temperature_t() : temp(0) {
+      last_decay = ceph_clock_now().tv.tv_sec;
+    }
+    temperature_t(uint32_t t) {
+      temp = MIN(t, TEMPMAX);
+      last_decay = ceph_clock_now().tv.tv_sec;
+    }
+    temperature_t(uint32_t t, uint32_t l) : last_decay(l) {
+      temp = MIN(t, TEMPMAX);
+    }
+    temperature_t(const temperature_t& ti) : last_decay(ti.last_decay) {
+      temp = MIN(ti.temp, TEMPMAX);
+    }
+
+    DENC(temperature_t, v, p) {
+      DENC_START(1, 1, p);
+      denc_varint(v.temp, p);
+      denc(v.last_decay, p);
+      DENC_FINISH(p);
+    }
+  };
+
+  class RankHistogram {
+    std::vector<int32_t> h; // histogram
+    uint32_t total;
+
+    private:
+      void _expand_to(unsigned s) {
+	if (s > h.size())
+	  h.resize(s, 0);
+      }
+      void _contract() {
+	unsigned p = h.size();
+	while (p > 1 && h[p-1] == 0)
+	  --p;
+	h.resize(p);
+      }
+    public:
+      RankHistogram() : total(0) {}
+
+      void clear() {
+	h.clear();
+        total = 0;
+      }
+
+      void add(uint32_t v) {
+	unsigned bin = cbits(v);
+	bin = MIN(31, bin);
+	_expand_to(bin + 1);
+	h[bin]++;
+	total++;
+	_contract();
+      }
+
+      void insert(uint32_t v) {
+	unsigned bin = cbits(v);
+	if (bin > 0 && bin != cbits(v-1)) {
+	  _expand_to(bin + 1);
+	  h[bin]++;
+	  h[bin-1]--;
+	  _contract();
+	}
+      }
+
+      void sub(uint32_t v) {
+	unsigned bin = cbits(v);
+	if (total > 0 &&
+      bin + 1 <= h.size() &&
+      h[bin] > 0) {
+	  h[bin]--;
+	  total--;
+	  _contract();
+	}
+      }
+
+      void decay(uint32_t bits) {
+	if (h.empty())
+	  return;
+	auto p = h.begin();
+	/// Objects at a temperature of 0 don't decay
+	p++;
+	for ( ; p != h.end(); ++p) {
+	  /// demote objects to a lower bar
+	  *(p-1) += *p - (*p >> bits);
+	  *p >>= bits;
+	}
+      }
+
+      uint32_t get_position_value(uint64_t pos) {
+        if (pos >= total)
+          return 0;
+	int64_t target = total - pos;
+	uint32_t bins = 1;
+	for (auto p = h.begin(); p != h.end(); ++p) {
+          target -= *p;
+	  if (target <= 0) {
+                break;
+          }
+	  bins <<= 1;
+	}
+	return bins;
+      }
+
+      uint32_t get_avg() const {
+	float avg = 0;
+	int bin = 0;
+	for (auto p = h.begin(); p != h.end(); ++p) {
+	  avg += (float)(*p) / total * bin;
+	  bin = (bin ? bin << 1 : 1);
+	}
+	return (uint32_t)avg;
+      }
+    };
+
+private:
+  typedef ceph::unordered_map<
+    uint32_t, temperature_t, ExclusiveKeyAsHash<unsigned>
+    > HitTable;
```

Comment:
I wonder if a better approach here would be to preallocated a single large vector for all of the hash values that fall into the PG's range.  This way the memory usage is more efficient (fully contiguous, no fragmentation) and also predictable (we see the full memory usage immediately, instead of having it grow slowly over time).

You could also have a precision option so that some number of bits are ignored, causing some additional overlap in objects with close hash values, but saving memory.